### PR TITLE
feat: fix coinbase

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The coinbase opcode doesn't give a Kakarot registered EVM address.

Resolves: #577 

## What is the new behavior?

- The coinbase opcode gives a Kakarot registered EVM address.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by bajpai244** [comment]
Created an [issue](https://github.com/kkrt-labs/kakarot-ssj/issues/588) to keep track of the TODO.

**Review by enitrat — COMMENTED** [review]
missing test for `exec_coinbase`

**Comment by enitrat** [comment]
unless you can't test it yet?

**Comment by bajpai244** [comment]
> unless you can't test it yet?

@enitrat , The current ability to deploy these sequencer accounts  on deploy of Kakarot is in TODO, I can setup a test without this functionality as well, I don't think should be an issue, but I don't think that will be a long term solution.

Lemme know, if an interim test works, then I can set one up.

**Comment by enitrat** [comment]
>  I can setup a test without this functionality as well, I don't think should be an issue, but I don't think that will be a long term solution.

I think the test of `exec_coinbase` should be decorrelated from the fact that deploying KKT also deploys predefined EOAs. You can deploy a simple EOA, set the sequencer address in a mock to this EOA's SN Address, and then test that coinbase has the right behavior

**Comment by Eikix** [comment]
> > I can setup a test without this functionality as well, I don't think should be an issue, but I don't think that will be a long term solution.
> 
> I think the _unit_ test of `exec_coinbase` should be decorrelated from the fact that deploying KKT also deploys predefined EOAs. You can deploy a simple EOA, set the sequencer address in a mock to this EOA's SN Address, and then test that coinbase has the right behavior

Yes the starknet::testing function exists
```rust
// Set the sequencer address to the provided value.
pub fn set_sequencer_address(address: ContractAddress) {
    cheatcode::<'set_sequencer_address'>(array![address.into()].span());
}
```

1. Deploy an EOA
2. Set the sequencer address to the EOA's stakrnet address
3. exec_coinbase
4. Check that the returned address is the EOA's

File: crates/evm/src/tests/test_instructions/test_block_information.cairo [hunk 1]
```diff
@@ -208,3 +214,27 @@ fn test_randao_should_push_zero_to_stack() {
...
+// *************************************************************************
+// 0x41: COINBASE
+// *************************************************************************
+#[test]
+#[available_gas(5000000)]
+fn test_exec_coinbase() {
+    // Given
+    let (native_token, kakarot_core) = setup_contracts_for_testing();
+    let sequencer_eoa = kakarot_core.deploy_eoa(evm_address());
+    fund_account_with_native_token(sequencer_eoa, native_token, 0x1);
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_block_information.cairo:227]: why do you need to fund the account?

**Comment by bajpai244** [comment]
@Eikix, @enitrat added the unit test + added some unwraps to this file where I could see we were missing, done via [this](https://github.com/kkrt-labs/kakarot-ssj/pull/581/commits/0155d09638405ec0bbf4e0e07bfdbab3a9a5c03d) commit.

File: crates/evm/src/tests/test_instructions/test_block_information.cairo (hunk 1 above)
Comment (bajpai244) [reply crates/evm/src/tests/test_instructions/test_block_information.cairo]: Ah, I copied the boilerplate for the test, and it got in as well, removed via [this](https://github.com/kkrt-labs/kakarot-ssj/pull/581/commits/3c7f2c3a365f130329be4be3c6f814af5f2318d0) commit.

**Pull request merged**